### PR TITLE
168 expected penalties are applied to grade after extra credit points are added

### LINKED ISSUE #168: Expected penalties are applied to grade after extra credit points are added
## Overview
Adjust the calculation of final scores so that any late penalties (or other penalties applied for low commit frequency etc) are applied based on a percentage of the final score _after_ any extra credit points are accounted for.

The grade should be calculated as the following formula:
```
LATE_PERCENTAGE_PER_DAY = 10
LOW_COMMIT_PENALTY_PCT = 10

gradedScore = percentage_tests_passed + extra_extra_credit_points_available
latePenaltyModifier = 1 - (LATE_PERCENTAGE_PER_DAY/100 * effective_days_late)
lowCommitModifier = 1 - (commits_under_permissible_threshold ? 0 : LOW_COMMIT_PENALTY_PCT)
finalScore = gradedScore * latePenaltyModifier * lowCommitModifier

return finalScore;
```

## Discussion
Professor Jensen said the following:
> I would have expected penalties (late, git commits) to apply after calculating the final score.

This was said in response to the following TA discussion
> **TA 1:** Should the late penalty apply to the extra credit points? It looks like the autograder is not currently doing that. For example, a student got 78% turning it in 3 days late, rather than 75.6% I would expect if the late penalty applied to those points

> **TA 2:** Personally I don't think it should penalize the extra credit, although I'd agree that the syllabus wording isn't perfectly clear: "The late penalty for all assignments is 10% off per school day"

> **TA 3:** This [proposed adjustment] would also be consistent with the behavior for partial credit submissions which may occur on phase 6.

##  Test Cases
* A student passes all the tests and the extra credit tests for a total raw score of **108%**. However, the student turns in the code **one school day** late (10% per day). This results in a final score of **97.2%** 
* On a later phase, a student receives credit for **60%** of the points on the rubric. However, the student only committed on **2 days** instead of the 3 separate days required by #160. Therefore, a 10% penalty is applied for a total score of **54%**.
* A student passes all of the test cases, but turns in the code **8 school days** late. He is only penalized for the first 5 of these, so the final score if **50%**.

### PR DESCRIPTION
**Pull request by frozenfrank**
This closes  #168.

It also begins to put some of the key functionality in place for #156 via a `GradingSettings` record.

While I have confirmed that this code compiles, I'm not sure how to run it on submissions to verify that the results are correct. Please do that, or help me figure out how to.

### CODE REVIEW

**Comment by frozenfrank** [comment]
@19mdavenport See my comment to Paul above. Maybe would you help me run the code in order to test it on submissions?

File: src/main/java/edu/byu/cs/autograder/Grader.java [hunk 1]
```diff
@@ -112,10 +112,18 @@ public Grader(String repoUrl, String netId, Observer observer, Phase phase) thro
...
         this.observer = observer;
     }
+    private GradingSettings getGradingSettings() {
+        // FIXME! Import from some dynamic location
+        return new GradingSettings(
+                5,
+                10,
+                10
+        );
+    }
```
Comment (pawlh) [inline src/main/java/edu/byu/cs/autograder/Grader.java:126]: I like the future prep here 💯

**Comment by frozenfrank** [comment]
That's fine, I'll tackle it. Thanks for letting me know.

**Comment by frozenfrank** [comment]
The underlying issue has already been resolved by #263.

If we want to preserve any of the other ideas from this PR, they'll come from a new one.

**Pull request closed without merging**